In [289]:
import numpy as np 
import pandas as pd
from matplotlib.pyplot import plot as plt
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
dataFrame = pd.read_csv('sonar_csv.csv')

In [2]:
dataFrame.isnull().sum().sum()

0

دیتا نال نداریم در نتیجه نیازی به حذف کردن هم نداریم 

In [275]:
dataFrame.head(5)

attribute_1  attribute_2  attribute_3  attribute_4  attribute_5  \
0       0.0200       0.0371       0.0428       0.0207       0.0954   
1       0.0453       0.0523       0.0843       0.0689       0.1183   
2       0.0262       0.0582       0.1099       0.1083       0.0974   
3       0.0100       0.0171       0.0623       0.0205       0.0205   
4       0.0762       0.0666       0.0481       0.0394       0.0590   

   attribute_6  attribute_7  attribute_8  attribute_9  attribute_10  ...  \
0       0.0986       0.1539       0.1601       0.3109        0.2111  ...   
1       0.2583       0.2156       0.3481       0.3337        0.2872  ...   
2       0.2280       0.2431       0.3771       0.5598        0.6194  ...   
3       0.0368       0.1098       0.1276       0.0598        0.1264  ...   
4       0.0649       0.1209       0.2467       0.3564        0.4459  ...   

   attribute_52  attribute_53  attribute_54  attribute_55  attribute_56  \
0        0.0027        0.0065        0.0159        0.0072        0.0167   
1        0.0084        0.0089        0.0048        0.0094        0.0191   
2        0.0232        0.0166        0.0095        0.0180        0.0244   
3        0.0121        0.0036        0.0150        0.0085        0.0073   
4        0.0031        0.0054        0.0105        0.0110        0.0015   

   attribute_57  attribute_58  attribute_59  attribute_60  Class  
0        0.0180        0.0084        0.0090        0.0032   Rock  
1        0.0140        0.0049        0.0052        0.0044   Rock  
2        0.0316        0.0164        0.0095        0.0078   Rock  
3        0.0050        0.0044        0.0040        0.0117   Rock  
4        0.0072        0.0048        0.0107        0.0094   Rock  

[5 rows x 61 columns]

دیتا را با استفاده از کم کردن از میانگین و تقسیم کردن بر واریانس نرمال میکنیم 

In [276]:
dataFrame_normal = (  dataFrame.iloc[:,:60] - dataFrame.iloc[:,:60].mean()  ) / dataFrame.iloc[:,:60].var()

دیتا را به صورت رندوم و با نسبت 80 به 20 تقسیم میکنیم 

In [10]:
inds = np.random.choice(range(208), (208 // 5)*4 , replace=False)
mylist = []
for i in range(0,208):
    if i not in inds:
        mylist.append(i)  
        
X_train = dataFrame_normal.values[inds]
X_test = dataFrame_normal.values[mylist]
Y_train = dataFrame['Class'].values[inds]
Y_test = dataFrame['Class'].values[mylist]

0.8636363636363636

## Linear 

In [302]:
clf_linear = SVC(kernel='linear',C=0.5,class_weight='balanced')
clf_linear.fit(X_train, Y_train)
y_predict_test_linear = clf_linear.predict(X_test)
y_predict_train_linear = clf_linear.predict(X_train)
accuracy_test_linear = round( accuracy_score(Y_test , y_predict_test_linear) , 2 )
accuracy_train_linear = round(  accuracy_score(Y_train , y_predict_train_linear) , 2 )

## ploynomial

In [305]:
clf_poly = SVC(kernel='poly',C = 250,degree=2,coef0=25)
clf_poly.fit(X_train, Y_train)
y_predict_test_polynomial = clf_poly.predict(X_test)
y_predict_train_polynomial = clf_poly.predict(X_train)
accuracy_test_poly = round( accuracy_score(Y_test , y_predict_test_polynomial) , 2 )
accuracy_train_poly = round( accuracy_score(Y_train , y_predict_train_polynomial) , 2 )

## RBF

In [308]:
clf_rbf = SVC(kernel='rbf',C=40)
clf_rbf.fit(X_train, Y_train)
y_predict_test_rbf = clf_rbf.predict(X_test)
y_predict_train_rbf = clf_rbf.predict(X_train)
accuracy_test_rbf = round( accuracy_score(Y_test , y_predict_test_rbf) , 2 )
accuracy_train_rbf = round ( accuracy_score(Y_train , y_predict_train_rbf) , 2 )

In [309]:
print('accuracy train linear' , accuracy_train_linear ) 
print('accuracy test linear' , accuracy_test_linear )
print('accuracy train polynomial' , accuracy_train_poly )
print('accuracy test polynomial' , accuracy_test_poly )
print('accuracy train RBF ' , accuracy_train_rbf )
print('accuracy test RBF ' , accuracy_test_rbf )

accuracy train linear 0.99
accuracy test linear 0.86
accuracy train polynomial 1.0
accuracy test polynomial 0.91
accuracy train RBF  0.99
accuracy test RBF  0.77


| SVM kernel  | Kernel parameter Value | Accuracy on Train | Accuracy on Test |
| --- | --- | --- | --- |
| Linear | C=0.5 | 0.99 | 0.86 | 
| polynomial | C=250 , degree=2 , coef0=25 | 1 | 0.91 |
| RBF | C=40 | 0.99 | 0.77 | 